core/src/main/java/com/linecorp/armeria/internal/common/TimeoutScheduler.java

```diff
@@ -416,13 +435,21 @@ private void invokeTimeoutTask() {
             if (!whenTimedOutUpdater.compareAndSet(this, null, COMPLETED_FUTURE)) {
                 whenTimedOut.doComplete();
             }
+        } else {
+            // Set CANCELLED flag first to prevent duplicate execution
+            state = State.CANCELLED;
```

Comment:
Don't we need to complete `whenTimingOut` and `whenTimedOut` even if it is cancelled. Unless the futures never complete forever.

Reply:
How about calling `cancel()` the futures?

Reply:
That sounds nice. Could you reuse cacelled futures if users don’t call `whenTime*out` yet to avoid additional memory allocation?

Reply:
Aren't the futures `null` when users haven't called `whenTim*out`?

Reply:
Yeah, what I want to say was 😀:
```
if (!isTimedOut) {
	if (!whenTimedOutUpdater.compareAndSet(this, null, CANCELLED_FUTURE)) {
	    whenTimedOut.doCancel();
	}
}
```

Reply:
Oh sure 😁